docs/designs/custom_allreduce_training_loop_support.md

```diff
@@ -0,0 +1,402 @@
+# Support Customizing the Training Loop Using AllReduce
+
+This document describes the design for supporting customizing
+the training loop when users use the AllReduce distribution strategy.
+
+## Motivation
+
+Users need to define a forward computation, a loss function, an optimizer,
+and a dataset function in ElasticDL. ElasticDL provides
+the training loop with those definitions. ElasticDL only supports
+Keras API to define the forward computation. This is not very flexible for
+users to define some complex models in CV or NLP. Also, users may wan to
+customize the training loop to control the model iteration.
+For example, given a pre-trained network, users might only want to
+optimize a new set of output weights.
+
+If ElasticDL supports customizing the training loop, users can use different
+deep-learning libraries (e.g., TensorFlow, Pytorch) to define their training
+loop. ElasticDL only needs to provide dynamic data partitioning for
+dataset and elastic AllReduce to merge gradients across workers.
+
+## The Training Loop of ElasticDL
+
+The training loop of ElasticDL on the worker contains two steps:
+
+- Get tasks from the master to create dataset with data shards in tasks.
+- Execute training loop to calculate gradients and update the model
+using the gradients.
+
+![Training Loop](../images/training_loop.jpg)
+
+In order to support customized training loop with ElasticDL AllReduce,
+ElasticDL provides users with a dataset with dynamic partition support,
+and an API `elastic_run` for elastic training. Users need to implement
+their own training logic for a training data batch in a function
+`train_one_batch`.
+
+```Python
+def train_one_batch(batch_data):
+    loss = forward(batch_data)
+    gradients = backward(loss)
+    gradients = allreduce(gradients)
+    upate_model(gradients)
+```
+
+Then users can write the training loop using `train_one_batch`,
+the dataset and `elastic_run` API provided by ElasticDL.
+
+```python
+def train(dataset, elastic_manager):
+    """
+    Arguments:
+        dataset: tf.data.Dataset. ElasticDL creates a dataset according to
+            tasks from the master
+        elastic_manager: A manager to execute elastic training.
+    """
+    for batch_data in dataset:
+        elastic_train = elastic_manager.elastic_run(train_one_batch)
+        elastic_train(batch_data)
+```
+
+## Elastic AllReduce Wrapper for Customized Training Loop
+
+ElasticDL AllReduce is based on Horovod. Using Horovod, users only need to
+use Horovod APIs to wrap optimizer or tape to execute AllReduce in their
+training loop.
+
+```python
+tape = hvd.DistributedGradientTape(tape)
+grads = tape.gradient(loss, model.trainable_variables)
+optimizer.apply_gradients(
+    zip(grads, model.trainable_variables)
+)
+```
+
+We can develop a decorator to wrap the gradient combination process
```

Comment:
Thus, we can implement `elastic_run` as a decorator to wrap the gradient combination process using AllReduce.